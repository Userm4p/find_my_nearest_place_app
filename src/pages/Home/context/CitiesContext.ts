import { createContext } from "react";
import { City } from "../types/cities";

export interface ICitiesContext {
  cities: City[];
  values: {
    city: string;
  };
  handleChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  reset: () => void;
  citiesToTable: City[];
  handleNextPage: () => void;
  handlePrevPage: () => void;
  pagesNumber: number;
  page: number;
}

export const CitiesContext = createContext<ICitiesContext>({
  cities: [],
  values: {
    city: "",
  },
  handleChange: () => {},
  reset: () => {},
  citiesToTable: [],
  handleNextPage: () => {},
  handlePrevPage: () => {},
  pagesNumber: 0,
  page: 0,
});
