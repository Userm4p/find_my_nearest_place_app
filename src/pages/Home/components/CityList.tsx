import { useContext, useMemo } from "react";
import { CitiesContext } from "../context/CitiesContext";
import { CitiesToTableEmpty, CityItem, PageButton } from "./styled";
import { Span } from "../../../components/Span/styled";

export const CityList = () => {
  const {
    citiesToTable,
    pagesNumber,
    page,
    handleNextPage,
    handlePrevPage,
    handleSelectCity,
  } = useContext(CitiesContext);

  const isPrevButtonDisabled = useMemo(() => page === 0, [page]);
  const isNextButtonDisabled = useMemo(
    () => page + 1 === pagesNumber,
    [page, pagesNumber],
  );

  const citiesToTableIsEmpty = useMemo(
    () => citiesToTable.length === 0,
    [citiesToTable],
  );

  return (
    <div className="Home__CityList_container">
      <div className="Home__CityList_header">
        <div className="Home__CityList_header_pages_container">
          <Span>Page</Span>
          <Span>{page + 1}</Span>
          <Span>of</Span>
          <Span>{pagesNumber}</Span>
        </div>
        <div className="Home__CityList_header_buttons_container">
          <PageButton
            data-testid="prev_button"
            onClick={handlePrevPage}
            disabled={isPrevButtonDisabled}
          >
            Prev
          </PageButton>
          <PageButton
            data-testid="next_button"
            onClick={handleNextPage}
            disabled={isNextButtonDisabled}
          >
            Next
          </PageButton>
        </div>
      </div>
      <div className="Home__CityList_cities_container">
        {citiesToTable.map((city, i) => (
          <CityItem onClick={() => handleSelectCity(city)} key={i}>
            {city.name}
          </CityItem>
        ))}
        {citiesToTableIsEmpty && (
          <CitiesToTableEmpty data-testid="empty_cities_list">
            We couldn't find any city with the name you provided. ðŸ˜¢
          </CitiesToTableEmpty>
        )}
      </div>
    </div>
  );
};
