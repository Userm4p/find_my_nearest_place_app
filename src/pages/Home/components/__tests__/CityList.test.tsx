import { render } from "@testing-library/react";
import { CityList } from "../CityList";
import { CitiesContext } from "../../context/CitiesContext";
import "@testing-library/jest-dom";

const mockContext = {
  citiesToTable: Array.from({ length: 20 }, (_, i) => ({
    name: `City ${i}`,
    country: `Country ${i}`,
    lat: `${i}`,
    lng: `${i}`,
  })),
  pagesNumber: 20,
  page: 1,
  handleNextPage: () => {},
  handlePrevPage: () => {},
  values: { city: "" },
  handleChange: () => {},
  reset: () => {},
  cities: [],
  handleSelectCity: () => {},
  nearestCities: [],
  selectedCity: null,
  clearSelectedCity: () => {},
};

describe("Tests in CityList", () => {
  test("should match snapshot", () => {
    const { container } = render(
      <CitiesContext.Provider value={mockContext}>
        <CityList />
      </CitiesContext.Provider>,
    );
    expect(container).toMatchSnapshot();
  });
  test("prev button should be disabled", () => {
    const { getByTestId } = render(
      <CitiesContext.Provider value={{ ...mockContext, page: 0 }}>
        <CityList />
      </CitiesContext.Provider>,
    );
    expect(getByTestId("prev_button")).toBeDisabled();
  });

  test("next button should be disabled", () => {
    const { getByTestId } = render(
      <CitiesContext.Provider value={{ ...mockContext, page: 19 }}>
        <CityList />
      </CitiesContext.Provider>,
    );
    expect(getByTestId("next_button")).toBeDisabled();
  });

  test("on click city should call handleSelectCity", () => {
    const handleSelectCity = jest.fn();
    const { getByText } = render(
      <CitiesContext.Provider value={{ ...mockContext, handleSelectCity }}>
        <CityList />
      </CitiesContext.Provider>,
    );
    getByText("City 0").click();
    expect(handleSelectCity).toHaveBeenCalledWith(mockContext.citiesToTable[0]);
  });

  test("should show empty message", () => {
    const { getByTestId } = render(
      <CitiesContext.Provider value={{ ...mockContext, citiesToTable: [] }}>
        <CityList />
      </CitiesContext.Provider>,
    );
    expect(getByTestId("empty_cities_list")).toBeInTheDocument();
  });
});
