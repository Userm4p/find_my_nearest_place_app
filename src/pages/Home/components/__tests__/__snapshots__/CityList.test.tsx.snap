// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in CityList should match snapshot 1`] = `
<div>
  <div
    class="Home__CityList_container"
  >
    <div
      class="Home__CityList_header"
    >
      <div
        class="Home__CityList_header_pages_container"
      >
        <span
          class="sc-jTQCzO gJpYtT"
        >
          Page
        </span>
        <span
          class="sc-jTQCzO gJpYtT"
        >
          2
        </span>
        <span
          class="sc-jTQCzO gJpYtT"
        >
          of
        </span>
        <span
          class="sc-jTQCzO gJpYtT"
        >
          20
        </span>
      </div>
      <div
        class="Home__CityList_header_buttons_container"
      >
        <button
          class="sc-guDLey csPTfk"
          data-testid="prev_button"
        >
          Prev
        </button>
        <button
          class="sc-guDLey csPTfk"
          data-testid="next_button"
        >
          Next
        </button>
      </div>
    </div>
    <div
      class="Home__CityList_cities_container"
    >
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 0
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 1
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 2
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 3
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 4
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 5
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 6
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 7
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 8
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 9
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 10
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 11
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 12
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 13
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 14
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 15
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 16
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 17
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 18
      </div>
      <div
        class="sc-beySPh hWTDYJ"
      >
        City 19
      </div>
    </div>
  </div>
</div>
`;
